Improving Decoding Capability accuracy on Media Capability API

youennf opened the pull request:
#### 044bcbf0c269e999b8bcda26e7de11041e057c0d
<pre>
Improving Decoding Capability accuracy on Media Capability API
<a href="https://rdar.apple.com/150685443">rdar://150685443</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=292468">https://bugs.webkit.org/show_bug.cgi?id=292468</a>

Reviewed by Jean-Yves Avenard.

The regular streaming media capabilities code path is doing finer grained checks on things like HDR.
We reuse that code path for WebRTC video decoding media capabilities by first checking whether the video decoder
is supported by WebRTC and is power efficient.
If so, we make another call to validate that the same video configuration would be supported for MSE.
We then reuse these results.

The reason we do this two step approach is that, HW support be the same for WebRTC and regular video streaming.
But SW support may be different (VP8 support for instance).

We update GStreamerRegistryScanner::isRtpPacketizerSupported to add missing FourCC based codec names (vp08, vp09, avc1....).

* LayoutTests/webrtc/media-capabilities-webrtc-expected.txt: Added.
* LayoutTests/webrtc/media-capabilities-webrtc.html: Added.
* Source/WebCore/platform/graphics/AV1Utilities.cpp:
(WebCore::parseEnumFromStringView): Deleted.
* Source/WebCore/platform/graphics/AV1Utilities.h:
(WebCore::parseEnumFromStringView):
* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
(WebCore::GStreamerRegistryScanner::isRtpPacketizerSupported):
* Source/WebCore/platform/mediastream/WebRTCProvider.cpp:
(WebCore::contentTypeFromRTPVideoMimeType):
(WebCore::WebRTCProvider::createDecodingConfiguration):
* Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCProvider.cpp:
(WebCore::LibWebRTCProvider::videoDecodingCapabilitiesOverride):

Canonical link: <a href="https://commits.webkit.org/295001@main">https://commits.webkit.org/295001@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2885ba77ee869a9f20b21c75d80ac623000f783a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103791 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23493 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13813 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108983 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54443 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105831 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23844 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32039 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78853 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106797 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18506 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93631 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59187 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18311 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11680 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53819 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/88081 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11737 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111371 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30947 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/22812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87855 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31309 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89832 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87509 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22275 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32398 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10120 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25304 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30875 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36178 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30669 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34004 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32230 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/platform/mediastream/WebRTCProvider.cpp:
```diff
@@ -198,6 +200,107 @@ bool WebRTCProvider::isH264EncoderSmooth(const VideoConfiguration&)
+    }
+
+    if (equalLettersIgnoringASCIICase(containerType, "video/h265"_s)) {
+        auto profileSpace = contentType.parameter("profile-space"_s);
```
The contents of this `if` statement should use `HEVCUtilities.h` to create the `codec` string by building a `HEVCParameters` object.

in fact, the parsing portion should likely also be added to `HEVCUtilities.h`.

youennf replied:
Sounds good, done

youennf replied:
As for the parsing portion, this is RTP specific, so I am not sure it is good to move them to the utilities.h files.

jernoble commented on Source/WebCore/platform/mediastream/WebRTCProvider.cpp:
```diff
@@ -198,6 +200,107 @@ bool WebRTCProvider::isH264EncoderSmooth(const VideoConfiguration&)
+    }
+
+    if (equalLettersIgnoringASCIICase(containerType, "video/vp9"_s)) {
+        auto profileId = contentType.parameter("profile-id"_s);
```
Ditto for this function and `VP9UtilitiesCocoa.h`.

jernoble commented on Source/WebCore/platform/mediastream/WebRTCProvider.cpp:
```diff
@@ -198,6 +200,107 @@ bool WebRTCProvider::isH264EncoderSmooth(const VideoConfiguration&)
+    }
+
+    if (equalLettersIgnoringASCIICase(containerType, "video/av1"_s)) {
+        auto profile = contentType.parameter("profile"_s);
```
Ditto for `AV1UtilitiesCocoa.h`.

jernoble commented on Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCProvider.cpp:
```diff
@@ -528,13 +528,12 @@ std::optional<MediaCapabilitiesDecodingInfo> LibWebRTCProvider::videoDecodingCap
         info.powerEfficient = info.smooth = true;
     else if (equalLettersIgnoringASCIICase(containerType, "video/av1"_s))
-        info.powerEfficient = false;
+        info.powerEfficient = info.smooth = true;
```
The answers to `powerEfficient` seem incorrect here?

youennf replied:
powerEfficient should be true if AV1 is only exposed when there is HW AV1 support.
That said, I reverted this change, we can tackle it in a follow-up.

jyavenard commented on Source/WebCore/platform/mediastream/WebRTCProvider.cpp:
```diff
@@ -198,6 +203,131 @@ bool WebRTCProvider::isH264EncoderSmooth(const VideoConfiguration&)
+
+        auto levelId = contentType.parameter("level-id"_s);
+        if (!levelId.isEmpty()) {
+            auto result = parseInteger<uint16_t>(levelId);
```
```suggestion
            if (auto result = parseInteger<uint16_t>(levelId))
                parameters.generalLevelIDC = *result;
            else
                return { };
                
```

youennf replied:
We usually prefer returning early in error case.

jyavenard commented on Source/WebCore/platform/mediastream/WebRTCProvider.cpp:
```diff
@@ -198,6 +203,131 @@ bool WebRTCProvider::isH264EncoderSmooth(const VideoConfiguration&)
+    }
+
+    if (equalLettersIgnoringASCIICase(containerType, "video/vp8"_s))
+        return "video/webm;codecs=vp08.00.50.08"_s;
```
I don't believe we support extended codec parameter. so `"video/webm; codec=vp08"` is probably more appropriate

youennf replied:
I can see `isVP8CodecConfigurationRecordSupported` which checks profile, chrome and bit depth.
In practice, this does not really matter given VP8 is SW really.

jyavenard commented on Source/WebCore/platform/mediastream/WebRTCProvider.cpp:
```diff
@@ -231,6 +361,24 @@ void WebRTCProvider::createDecodingConfiguration(MediaDecodingConfiguration&& co
+            info.smooth = result.smooth;
+            info.powerEfficient = result.powerEfficient;
+            if (!info.supported)
+                info.supportedConfiguration = { };
```
let's rename this to `configuration`

youennf replied:
Ditto

jyavenard commented on Source/WebCore/platform/mediastream/WebRTCProvider.cpp:
```diff
@@ -231,6 +361,24 @@ void WebRTCProvider::createDecodingConfiguration(MediaDecodingConfiguration&& co
+    if (info.powerEfficient && info.supportedConfiguration.video) {
+        auto videoConfiguration = info.supportedConfiguration;
+        videoConfiguration.audio = { };
+        videoConfiguration.video->contentType = contentTypeFromRTPVideoMimeType(info.supportedConfiguration.video->contentType);
```
let's rename `supportedConfiguration` to `configuration` or in. follow-up change

youennf replied:
It will be done in a separate PR where we change the IDL as well.

jyavenard commented on Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCProvider.cpp:
```diff
@@ -528,13 +528,14 @@ std::optional<MediaCapabilitiesDecodingInfo> LibWebRTCProvider::videoDecodingCap
+    else if (equalLettersIgnoringASCIICase(containerType, "video/h265"_s))
+        info.powerEfficient = info.smooth = true;
+    else if (equalLettersIgnoringASCIICase(containerType, "video/av1"_s)) {
+        // FIXME: Set value to true if AV1 is only enabled when HW decoder support is enabled.
```
I believe this is always the case, we only support av1 if we have a HW decoder available for playback
so we will return a false positive here.

youennf replied:
Yes, we should set `info.powerEfficient` to `true`. I'll do that as a follow-up since this is somehow orthogonal and requires test improvements.

youennf committed:
Improving Decoding Capability accuracy on Media Capability API
rdar://150685443
https://bugs.webkit.org/show_bug.cgi?id=292468

Reviewed by Jean-Yves Avenard.

The regular streaming media capabilities code path is doing finer grained checks on things like HDR.
We reuse that code path for WebRTC video decoding media capabilities by first checking whether the video decoder
is supported by WebRTC and is power efficient.
If so, we make another call to validate that the same video configuration would be supported for MSE.
We then reuse these results.

The reason we do this two step approach is that, HW support be the same for WebRTC and regular video streaming.
But SW support may be different (VP8 support for instance).

We update GStreamerRegistryScanner::isRtpPacketizerSupported to add missing FourCC based codec names (vp08, vp09, avc1....).

* LayoutTests/webrtc/media-capabilities-webrtc-expected.txt: Added.
* LayoutTests/webrtc/media-capabilities-webrtc.html: Added.
* Source/WebCore/platform/graphics/AV1Utilities.cpp:
(WebCore::parseEnumFromStringView): Deleted.
* Source/WebCore/platform/graphics/AV1Utilities.h:
(WebCore::parseEnumFromStringView):
* Source/WebCore/platform/graphics/gstreamer/GStreamerRegistryScanner.cpp:
(WebCore::GStreamerRegistryScanner::isRtpPacketizerSupported):
* Source/WebCore/platform/mediastream/WebRTCProvider.cpp:
(WebCore::contentTypeFromRTPVideoMimeType):
(WebCore::WebRTCProvider::createDecodingConfiguration):
* Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCProvider.cpp:
(WebCore::LibWebRTCProvider::videoDecodingCapabilitiesOverride):

Canonical link: https://commits.webkit.org/295001@main
